crates/eth-rpc/src/main.rs

```diff
@@ -15,12 +27,45 @@ async fn main() -> Result<()> {
     tracing_subscriber::FmtSubscriber::builder().with_env_filter(filter).finish().try_init()?;
 
     let starknet_config = StarknetConfig::from_env()?;
+
     let rpc_config = RPCConfig::from_env()?;
 
-    let provider = JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build();
-    let kakarot_client = KakarotClient::new(starknet_config, provider);
+    let provider: KakarotProvider = match &starknet_config.network {
+        Network::Madara | Network::Katana => {
+            KakarotProvider::JsonRpcClient(JsonRpcClientBuilder::with_http(&starknet_config).unwrap().build())
+        }
+
+        Network::Goerli1 => {
+            KakarotProvider::SequencerGatewayProvider(SequencerGatewayProvider::starknet_alpha_goerli())
+        }
+
+        Network::Goerli2 => {
+            KakarotProvider::SequencerGatewayProvider(SequencerGatewayProvider::starknet_alpha_goerli_2())
+        }
+
+        Network::Mainnet => {
+            KakarotProvider::SequencerGatewayProvider(SequencerGatewayProvider::starknet_alpha_mainnet())
+        }
+
+        Network::ProviderUrl(url) => {
+            KakarotProvider::JsonRpcClient(JsonRpcClientBuilder::new(HttpTransport::new(url.clone())).build())
+        }
+    };
+
+    let kakarot_client = match provider {
+        KakarotProvider::JsonRpcClient(provider) => {
+            KakarotClientType::RpcClient(KakarotClient::new(starknet_config, provider))
+        }
+        KakarotProvider::SequencerGatewayProvider(provider) => {
+            KakarotClientType::GatewayClient(KakarotClient::new(starknet_config, provider))
+        }
+    };
+
+    let (server_addr, server_handle) = match kakarot_client {
+        KakarotClientType::GatewayClient(kakarot_client) => run_server(Box::new(kakarot_client), rpc_config).await?,
```

Comment:
why is the `match` required? can't we use directly `run_server(Box::new(kakarot_client), rpc_config).await?`?

Reply:
because run_server takes in a kakarot_client type, not an enum wrapping this type

Reply:
we could implement on the enum an .into_inner() function that returns the inner type, using a match

Reply:
I agree! I just don't know yet how to do it

Reply:
Let's open an issue!